README.md
[english version](../../wiki/english)


**Kodi** ist ein kostenloser Open Source (GPL) Software Mediaplayer zum Verwalten und Abspielen von digitalen Medien. Kodi läuft unter Linux, OSX, Windows, iOS und Android. Wenn Du Kodi noch nicht kennst, besuche [kodi.tv](https://kodi.tv/) für mehr Informationen.

**S04tv** ist das Online Video Portal des FC Schalke 04. Jedes Spiel des S04 kann auf https://schalke04.de/tv/ re-Live in voller Länge angesehen werden. In der Regel sind die Spiele 15 Minuten nach Spielende verfügbar. Zusätzlich gibt es News (Montags bis Freitags ab 18 Uhr, außer an Spieltagen), Interviews und Sonderberichte. Viele Berichte sind kostenloas verfügbar, für die meisten Berichte (insbesondere die re-Live Spiele) ist jedoch ein kostenpflichtiger Benutzer-Account notwendig. Einige Videos sind für Schalke-Mitglieder kostenlos verfügbar, exklusive Inhalte benötigen ein kostenpflichtiges Abo (19,04 € für 12 Monate).


**Installation via Repo Browser**
  * In Kodi gehe zu: System -> Add-ons -> Get Add-ons -> Kodi Add-ons -> Video Add-ons --> S04tv
  * Klicke auf Install


**Installation via Datei**
  * Die neueste Installations-Datei kann hier runtergeladen werden: [link](https://github.com/maloep/xbmc-s04tv/releases)
  * In Kodi gehe zu System -> Add-ons -> Install from zip file
  * Navigiere zu der runtergeladenen Installations-Datei und klicke OK


**Verwendung**
  * Ohne gültigen Account kannst Du nur die Videos ansehen, die mit [FREE] gekennzeichnet sind. Für alle Videos, die mit [EXCL] gekennzeichnet sind, benötigst Du einen gültigen Account bei S04.tv.
  * Deine Account-Daten kannst Du in den Plugin-Einstellungen konfigurieren. Gehe zu Videos -> Add-ons, markiere S04tv und wähle "Menu" auf der Fernbedienung (oder "C" auf der Tastatur) um das Kontextmenu aufzurufen. Wähle "Add-on settings" und gib Deinen Benutzernamen und das Passwort ein. Beim nächsten Aufruf kannst Du alle Videos ansehen.
  * In den Plugin-Einstellungen kannst Du auch die Videoqualität einstellen und auswählen, ob Du das [FREE/EXCL]-Flag und/oder das Datum hinter dem Videotitel sehen möchtest und ob exklusive Beiträge ganz ausgeblendet werden sollen.

changelog.txt
3.2.1 (2020/01)
- update due to website changes

3.2.0 (2018/09)
- update due to website changes
- choose first or second half for re-live matches

3.1.7 (2018/08)
- marked as broken for Gotham

3.1.6 (2018/03)
- update due to website changes

3.1.5 (2018/02)
- update due to website changes

3.1.4 (2018/02)
- update due to website changes

3.1.3 (2017/11)
- update due to website changes

3.1.2 (2017/03)
- update due to website changes

3.1.1 (2017/03)
- update due to website changes

3.1.0 (2016/08)
- update due to website changes

3.0.2 (2015/12)
- update due to website changes
- play free content via youtube plugin

3.0.1 (2014/08)
- update due to website changes
- paging support

3.0.0 (2013/08)
- update due to website changes
- list videos in categories
- support of free and exclusive videos
- new options (video quality etc.)

2.0.1 (2013/07):
- another encoding fix

2.0.0 (2012/12):
- url encoding fixed

1.0.1 (2012/11):
- update due to website changes

1.0.0 (2012/02):
- Eden version bump

0.4.1 (2011/09):
- add logo to video items
- bugfix: title can be None

0.4 (2011/09):
- initial version

  

default.py
# -*- coding: utf-8 -*-

# Copyright (C) 2015 Malte Loepmann (maloep@googlemail.com)
#
# This program is free software; you can redistribute it and/or modify it under the terms
# of the GNU General Public License as published by the Free Software Foundation;
# either version 2 of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
# without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with this program;
# if not, see <http://www.gnu.org/licenses/>.

import xbmc, xbmcplugin, xbmcgui, xbmcaddon
import os, sys, re, json
import urllib
import uuid
from urlparse import parse_qs, urlparse
#import xml.etree.ElementTree as ET
import cookielib
try:
    # Python 2.6-2.7
    from HTMLParser import HTMLParser
except ImportError:
    # Python 3
    from html.parser import HTMLParser

from BeautifulSoup import BeautifulSoup
import mechanize

PLUGINNAME = 'S04tv'
PLUGINID = 'plugin.video.s04tv'
BASE_URL = 'https://schalke04.de/tv/videos/'

# Shared resources
addonPath = ''
addon = xbmcaddon.Addon(id='plugin.video.s04tv')
addonPath = addon.getAddonInfo('path')

BASE_RESOURCE_PATH = os.path.join(addonPath, "resources" )
sys.path.append( os.path.join( BASE_RESOURCE_PATH, "lib" ) )
sys.path.append( os.path.join( BASE_RESOURCE_PATH, "lib", "BeautifulSoup" ) )


language = addon.getLocalizedString
thisPlugin = int(sys.argv[1])
browser = mechanize.Browser()

missingelementtext = "Missing element '%s'. Maybe the site structure has changed."

videoquality = 'hd720'

quality = addon.getSetting('videoquality')
if quality == 'mid':
    videoquality = 'medium'
elif quality == 'high':
    videoquality = 'hd720'
elif quality == 'hd':
    videoquality = 'hd1080'



def buildVideoDir(url, doc):

    #allow sorting of video titles
    xbmcplugin.addSortMethod(thisPlugin, xbmcplugin.SORT_METHOD_UNSORTED)
    #xbmcplugin.addSortMethod(thisPlugin, xbmcplugin.SORT_METHOD_DATE)
    xbmcplugin.addSortMethod(thisPlugin, xbmcplugin.SORT_METHOD_TITLE)

    hideexclusive = addon.getSetting('hideexclusivevideos').upper() == 'TRUE'
    hideflag = addon.getSetting('hidefreeexclflag').upper() == 'TRUE'
    
    soup = BeautifulSoup(''.join(doc))
    section = soup.find('section')
    if(not section):
        xbmc.log(missingelementtext%'section')
        return
    
    title = ''
    url = ''
    imageUrl = ''
        
    ahrefs = section.findAll('a')
    
    for ahref in ahrefs:
        
        url = ahref['href']
        try:
            isPayedContent = ahref['data-payed'] == 'true'
        except KeyError:
            isPayedContent = False
        
        img = ahref.find('img')
        if(img):
            imageUrl = img['data-lazy-src']
               
        h3 = ahref.find('h3')
        mode = 2
        if(h3 is not None):
            title = ''
            parts = str(h3).decode('utf-8').split()
            for part in parts:
                title = title +' ' +part
            
            title = title.replace('<h3 class=\"card-title teaser-title\">', '').replace('</h3>', '').replace('<br />', '')
            title = HTMLParser().unescape(title)
        else:
            mode = 1
            title = language(30003)
        
        extraInfo = {}
        if not isPayedContent:
            extraInfo['isPayedContent'] = 'False'
        else:
            extraInfo['isPayedContent'] = 'True'
            if(hideexclusive):
                #don't add exclusive videos to list
                continue
            if(not hideflag):
                title = '[EXCL] ' +title
        
        if(mode == 1):
            addDir(title, url, mode, imageUrl)
        else:
            addLink(title, url, mode, imageUrl, '', extraInfo)


def getVideoUrl(url, doc):
    xbmc.log('getVideoUrl: url=' +url)

    is_logged_in, has_schalke_tv = login()
    if not is_logged_in:
        return

    isPayedContent = xbmc.getInfoLabel( "ListItem.Property(isPayedContent)" ) == 'True'
    #payedContent requires payed Schalke TV subscription
    if(isPayedContent):
        if not has_schalke_tv:
            xbmc.log("This video requires payed Schalke TV subscription")
            xbmcgui.Dialog().ok(PLUGINNAME, language(30104), language(30105))
            return
    
    #HACK: Free content may be hosted on youtube
    if(url.startswith("https://youtu.be/")):
        videoId = url.replace("https://youtu.be/", "")
        url='plugin://plugin.video.youtube/play/?video_id=' +videoId

        listitem = xbmcgui.ListItem(path=url)
        return xbmcplugin.setResolvedUrl(thisPlugin, True, listitem)

    #find the url to the dataoptions markup
    match_json_url = re.compile("url: '(.+)' \+").findall(doc)
    if not match_json_url:
        xbmc.log(missingelementtext % 'url: ')
        return

    json_url = 'https://schalke04.de' +match_json_url[0]+str(uuid.uuid1())
    response = getUrl(json_url)

    result = re.compile('data-id.+"(.+)",.+"container').findall(response)
    dataId = result[0].replace("\\", "")

    playoutUrl = 'https://playout.3qsdn.com/' \
                 + dataId + '?js=true&skin=s04&data-id=' \
                 + dataId + '&container=sdnPlayer_player&preview=false&width=100%25&height=100%25'

    response = getUrl(playoutUrl)

    match_playlist = re.compile('playlist: \((.+?)\)', re.DOTALL).findall(response)
    playlist = match_playlist[0]

    quote_keys_regex = r'([\{\s,])(\w+)(:)'
    playlist = re.sub(quote_keys_regex, r'\1"\2"\3', playlist)

    playlist = playlist.replace("'", '"')
    playlist = playlist.replace("\\x2F", '')

    videourl = ''

    jsonPlaylist = json.loads(playlist)
    for key in jsonPlaylist:
        entry = jsonPlaylist[key]
        quality = entry['quality']
        if (quality == videoquality):
            videourl = entry['src']

    listitem = xbmcgui.ListItem(path=videourl)
    return xbmcplugin.setResolvedUrl(thisPlugin, True, listitem)


def addDir(name, url, mode, iconimage):
    parameters = {'url' : url.encode('utf-8'), 'mode' : str(mode), 'name' : name.encode('utf-8')}
    u = sys.argv[0] +'?' +urllib.urlencode(parameters)
    ok = True
    listitem = xbmcgui.ListItem(name)
    listitem.setArt({'icon': 'DefaultFolder.png', 'thumb': iconimage})
    listitem.setInfo(type="Video", infoLabels={"Title": name})
    ok = xbmcplugin.addDirectoryItem(thisPlugin, u, listitem, isFolder=True)
    return ok
   

def addLink(name, url, mode, iconimage, date, extraInfo):
    parameters = {'url' : url.encode('utf-8'), 'mode' : str(mode), 'name' : name.encode('utf-8')}
    u = sys.argv[0] +'?' +urllib.urlencode(parameters)
    ok = True
    listitem = xbmcgui.ListItem(name)
    listitem.setArt({'icon': 'DefaultVideo.png', 'thumb': iconimage})
    if(date != ''):
        listitem.setInfo(type="Video", infoLabels={"Title": name, "Date": date})
    else:
        listitem.setInfo(type="Video", infoLabels={"Title": name})
    listitem.setProperty('IsPlayable', 'true')
    for key in extraInfo.keys():
        listitem.setProperty(key, extraInfo[key])
    ok = xbmcplugin.addDirectoryItem(thisPlugin, u, listitem)
    
    return ok


def login():

    username = addon.getSetting('username')
    xbmc.log('Logging in with username "%s"' %username)
    password = addon.getSetting('password')
    
    if(not username or not password):
        xbmcgui.Dialog().ok(PLUGINNAME, language(30102), language(30103))
        return False, False
    
    url = 'https://schalke04.de/account/login/'
    
    cj = cookielib.CookieJar()
    br = mechanize.Browser()
    br.set_cookiejar(cj)
    br.set_handle_robots(False)
    br.addheaders = [('User-agent',
                           'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:71.0) Gecko/20100101 Firefox/71.0')]
    br.open(url)

    for form in br.forms():
        try:
            form['email'] = username
            form['password'] = password
            br.form = form
            xbmc.log('Successfully set credentials in form')
            break
        except mechanize.ControlNotFoundError:
            pass

    xbmc.log('Submit login request')
    br.submit()

    br.open('https://schalke04.de/account/profil/')
    response = br.response().read()

    is_logged_in = '<li class="checked">First login' in response
    has_schalke_tv = '<li class="checked" data-account-placeholder="is_tv_subscriber">' in response

    xbmc.log('User is logged in = ' +str(is_logged_in))
    xbmc.log('User has Schalke TV Abo = ' + str(has_schalke_tv))

    return is_logged_in, has_schalke_tv


def getUrl(url):
        url = url.replace('&amp;','&')
        url = url.replace('&#38;','&')
        xbmc.log('Get url: '+url)
        browser.set_handle_robots(False)
        browser.addheaders = [('User-agent',
                          'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:71.0) Gecko/20100101 Firefox/71.0')]
        try:
            browser.open(url)
            response = browser.response().read()
        except Exception as exc:
            xbmc.log('Error while opening url: ' +str(exc))
            return ''
        return response


def runPlugin(url, doc):
    
    if mode is None or doc is None or len(doc)<1:
        buildVideoDir(url, doc)
       
    elif mode==1:
        buildVideoDir(url, doc)

    elif mode == 2:
        getVideoUrl(url, doc)


xbmc.log('S04TV: start addon')

params = parse_qs(urlparse(sys.argv[2]).query)
url = None
name = None
mode = None

try:
    url = params["url"][0]
except (IndexError, KeyError):
    xbmc.log("No parameter url found")
try:
    name = params["name"][0]
except (IndexError, KeyError):
    xbmc.log("No parameter name found")
try:
    mode = int(params["mode"][0])
except (IndexError, KeyError):
    xbmc.log("No parameter mode found")

xbmc.log("Mode: " +str(mode))
xbmc.log("URL: "+str(url))
xbmc.log("Name: "+str(name))

if url is None:
    url = BASE_URL

doc = getUrl(url)
runPlugin(url, doc)
xbmcplugin.endOfDirectory(thisPlugin)


tags/0.4.0/temptests.py
import os, sys
import urllib, urllib2, cookielib

# Shared resources
addonPath = ''
try:
	import xbmcaddon
	addon = xbmcaddon.Addon(id='plugin.video.s04tv')
	addonPath = addon.getAddonInfo('path')
except:
	addonPath = os.getcwd()

BASE_RESOURCE_PATH = os.path.join(addonPath, "resources" )
sys.path.append( os.path.join( BASE_RESOURCE_PATH, "lib" ) )
sys.path.append( os.path.join( BASE_RESOURCE_PATH, "lib", "BeautifulSoup" ) )
sys.path.append( os.path.join( BASE_RESOURCE_PATH, "lib", "mechanize" ) )

from BeautifulSoup import *
import mechanize

browser = mechanize.Browser()

url = 'https://www.s04tv.de/index.php/s04tv-kostenlos.html'
browser.open(url)

doc = browser.response().read()
soup = BeautifulSoup(''.join(doc))

imageTag = soup.find('div', attrs={'class' : 'field Bild'})
link = imageTag.find('a')
linkValue = link['href']

newUrl = 'https://www.s04tv.de/' +linkValue
browser.open(newUrl)

doc = browser.response().read()
soup = BeautifulSoup(''.join(doc))

videoTag = soup.find('video')
if(videoTag):
	videoUrl = videoTag['src']
	
print videoUrl






Testseiten/testplugin.py

# -*- coding: iso-8859-15 -*-

import os, sys, re, json
import urllib, urllib2
from urlparse import *
import xml.etree.ElementTree as ET

BASE_RESOURCE_PATH = "C:\\Users\\lom\\AppData\\Roaming\\XBMC\\addons\\plugin.video.s04tv.dev\\resources"
sys.path.append( os.path.join( BASE_RESOURCE_PATH, "lib" ) )
sys.path.append( os.path.join( BASE_RESOURCE_PATH, "lib", "BeautifulSoup" ) )

from BeautifulSoup import BeautifulSoup
from BeautifulSoup import Tag
  

BASE_URL = 'http://www.s04.tv'


def buildHomeDir(url, doc):
    
    soup = BeautifulSoup(''.join(doc))
    
    nav = soup.find('nav')
    
    for navitem in nav.contents:
        #first tag is our ul
        if(type(navitem) == Tag):
            for ulitem in navitem.contents:
                if(type(ulitem) == Tag):
                    if(ulitem.name == 'li'):
                        a = ulitem.find('a')
                        print a.text
                        print a['href']
            break
        
        
def buildSubDir(url, doc):
    
    soup = BeautifulSoup(''.join(doc))
    
    nav = soup.find('ul', attrs={'class': 'contentnav'})
    div =  nav.find('div')
    ul = div.find('ul')
    for ulitem in ul.contents:
        if(type(ulitem) == Tag):
            if(ulitem.name == 'li'):
                a = ulitem.find('a')
                print a.text
                print a['href']
                

def buildSubSubDir(url, doc):
    print 'buildSubSubDir'
    print 'url = ' +url
    soup = BeautifulSoup(''.join(doc))
    
    #get pagenumber
    indexpage = url.find('page/')
    indexpage = indexpage + len('page/')
    indexminus = url.find('-', indexpage) 
    pagenumber = url[indexpage:indexminus]
    
    #check if we have corresponding sub menus
    li = soup.find('li', attrs={'class':'dm_%s'%pagenumber})
    if(li == None):
        buildVideoLinks(url, doc)
        return
    
    div =  li.find('div')
    ul = div.find('ul')
    for ulitem in ul.contents:
        if(type(ulitem) == Tag):
            if(ulitem.name == 'li'):
                a = ulitem.find('a')
                print a.text
                print a['href']
        


def buildVideoDir(url, doc):
    
    hideexclusive = False
    hideflag = False
    hidedate = False
    
    soup = BeautifulSoup(''.join(doc))
    articles = soup.findAll('article')
    if(not articles):
        return
    
    for article in articles:
                
        div = article.find('div')
        if(not div):
            continue
        
        flag = div['class']
        
        #for some reason findNextSibling does not work here
        p = div.findAllNext('p', limit=1)
        if(not p):
            date = ''
        else:
            date = p[0].text
                
        img = div.findAllNext('img', limit=1)
        if(not img):
            continue
        imageUrl = img[0]['src']
        
        #HACK: this is only required on home page
        h2 = img[0].findAllNext('h2', limit=1)
        if(h2):
            a = h2[0].findAllNext('a', limit=1)
            if(not a):
                continue
            url = a[0]['href']
            span = a[0].find('span')
        else:
            a = img[0].findAllNext('a', limit=1)
            if(not a):
                continue
            url = a[0]['href']
            span = a[0].find('span')
        
        if(not span):
            continue
        
        title = ''
        for text in span.contents:
            if(type(text) != Tag):
                if(title != ''):
                    title = title +': '
                title = title +text
                
        #only required for homescreen
        span2 = span.nextSibling
        if(span2):
            title = title +': ' +span2.text
                
        if(not hidedate):
            title = title + ' (%s)'%date
                
        extraInfo = {}
        if(flag == 'flag_free'):
            if(not hideflag):
                title = '[FREE] ' +title
            extraInfo['IsFreeContent'] = 'True'
        elif(flag == 'flag_excl'):
            if(hideexclusive):
                #don't add exclusive videos to list
                continue
            if(not hideflag):
                title = '[EXCL] ' +title
            extraInfo['IsFreeContent'] = 'False'
                
        
        url = BASE_URL + url
        print title
        print url
        print imageUrl
        print date
        print flag        
        
        #addDir(title, url, 4, imageUrl, date, extraInfo)

        
def getVideoUrl(url):
    print 'getVideoUrl'
    
    doc = getUrl(url)
    soup = BeautifulSoup(''.join(doc))
    
    p = soup.find('p', attrs={'class': 'breadcrumbs'})
    a = p.find('a')
    title = a['title']
    if(title == ''):
        title = 'Play video'
    
    div = soup.find('div', attrs={'class': 'videobox'})
    script = div.find('script', attrs={'type': 'text/javascript'})
    
    scripttext = script.next
    indexbegin = scripttext.find("videoid: '")
    indexbegin = indexbegin + len("videoid: '")
    indexend = scripttext.find("'", indexbegin)
    xmlurl = scripttext[indexbegin:indexend]
    
    #load xml file
    xmlstring = getUrl(BASE_URL +xmlurl)
    root = ET.fromstring(xmlstring)
    
    urlElement = root.find('invoke/url')
    
    xmlstring = getUrl(urlElement.text)
    root = ET.fromstring(xmlstring)
    metas = root.findall('head/meta')
    vid_base_url = ''
    for meta in metas:
        if( meta.attrib.get('name') == 'httpBase'):
            vid_base_url = meta.attrib.get('content')
            break
    
    quality = 'low400'
    quality = '_%s.mp4'%quality
    videos = root.findall('body/switch/video')
    print 'quality: ' +quality 
    
    for video in videos:
        src = video.attrib.get('src')
        print 'src: ' +src
        if(src.find(quality) > 0):
            break
        
    print 'found video = ' +src
            

def login():
    
    username = 'bla'
    password = 'bla'
    print 'Logging in with username "%s"' %username
    
    loginparams = {'username_field' : username.encode('utf-8'), 'password_field' : password.encode('utf-8')}
    loginurl = 'https://ssl.s04.tv/get_content.php?lang=TV&form=login&%s' %urllib.urlencode(loginparams)    
    print loginurl
    loginresponse = getUrl(loginurl)
    print 'login response: ' +loginresponse
    
    #loginresponse should look like this: ({"StatusCode":"1","stat":"OK","UserData":{"SessionID":"...","Firstname":"...","Lastname":"...","Username":"lom","hasAbo":1,"AboExpiry":"31.07.14"},"out":"<form>...</form>"});
    #remove (); from response
    jsonstring = loginresponse[1:len(loginresponse) -2]
    jsonResult = json.loads(jsonstring)
    if(jsonResult['stat'] == "OK"):
        userdata = jsonResult['UserData']
        if(userdata['hasAbo'] == 1):
            print 'login successful'
            return True
        else:
            print 'login failed'
            return False
    else:
        print 'login failed'
        return False


def getUrl(url):
        url = url.replace('&amp;','&')
        print 'Get url: '+url
        req = urllib2.Request(url)
        req.add_header('User-Agent', 'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-GB; rv:1.9.0.3) Gecko/2008092417 Firefox/3.0.3')
        response = urllib2.urlopen(req)
        link=response.read()
        response.close()
        return link




#login()

"""
url = BASE_URL
doc = getUrl(url)
buildHomeDir(url, doc)
"""

"""
url = "http://www.s04.tv/de/saison/page/4--4--.html"
doc = getUrl(url)
buildSubDir(url, doc)
"""

"""
url = "http://www.s04.tv/de/saison/stimmen-zum-spiel/page/53--4--.html"
doc = getUrl(url)
buildSubSubDir(url, doc)
"""

"""
#url = "http://www.s04.tv/de/saison/highlights/saison-2013/14/testspiele/page/332--4--.html"
url = "http://www.s04.tv"
doc = getUrl(url)
buildVideoDir(url, doc)
"""

"""
url = "http://www.s04.tv/de/saison/highlights/saison-2011/12/dfb-pokal/mp4migration/110531_pokallights1_neu/page/116---262-.html"
#url = "http://www.s04.tv/de/saison/highlights/saison-2013/14/testspiele/mp4/130729_lokleipzig_schalke_s04tv/page/335---323-.html"
doc = getUrl(url)
getVideoUrl(url)
"""

import re
line = 'changeVideoPage(1, 181)'
match_res=re.compile('changeVideoPage\(0,(.+)\)', re.DOTALL).findall(line)
res = (int)(match_res[0])
print('res = ' +str(res))